File: iOS/REMatching/REMatching/Http/Requester/RoomRequester.swift
//
//  RoomRequester.swift
//  REMatching
//
//  Created by Leapfrog-Software on 2018/07/16.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

struct RoomData {
    
    let id: String
    let isChecked: Bool
    let score: Int
    var review: Int
    var name: String
    var place: String
    var rent: Int
    var phone: String
    var email: String
    
    init?(data: Dictionary<String, Any>) {
        
        guard let roomId = data["id"] as? String else {
            return nil
        }
        self.id = roomId
        
        self.isChecked = (data["isChecked"] as? String) == "1"
        self.score = Int(data["score"] as? String ?? "0") ?? 0
        self.review = Int(data["review"] as? String ?? "0") ?? 0
        self.name = (data["name"] as? String)?.base64Decode() ?? ""
        self.place = (data["place"] as? String)?.base64Decode() ?? ""
        self.rent = Int(data["rent"] as? String ?? "0") ?? 0
        self.phone = (data["phone"] as? String)?.base64Decode() ?? ""
        self.email = (data["email"] as? String)?.base64Decode() ?? ""
    }
}

class RoomRequester {
    
    static let shared = RoomRequester()
    
    var dataList = [RoomData]()
    
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        let params = [
            "command": "getRoom"
        ]
        ApiManager.post(params: params) { [weak self] result, data in
            if result, let dic = data as? Dictionary<String, Any> {
                if dic["result"] as? String == "0", let datas = dic["data"] as? Array<Dictionary<String, Any>> {
                    self?.dataList = datas.compactMap { RoomData(data: $0) }
                    completion(true)
                    return
                }
            }
            completion(false)
        }
    }
    
    func checkedRooms() -> [RoomData] {
        return self.dataList.filter { $0.isChecked == true }
    }
    
    class func post(name: String, place: String, rent: Int, phone: String, email: String, completion: @escaping ((Bool, String?) -> ())) {
        
        let params = [
            "command": "postRoom",
            "name": name.base64Encode() ?? "",
            "place": place.base64Encode() ?? "",
            "rent": "\(rent)",
            "phone": phone.base64Encode() ?? "",
            "email": email.base64Encode() ?? ""
        ]
        ApiManager.post(params: params) { (result, data) in
            if result, let dic = data as? Dictionary<String, Any> {
                if dic["result"] as? String == "0", let roomId = dic["roomId"] as? String {
                    completion(true, roomId)
                    return
                }
            }
            completion(false, nil)            
        }
    }

}

File: Android/app/src/main/java/leapfrog_inc/rematching/Http/Requester/RoomRequester.java
package leapfrog_inc.rematching.Http.Requester;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.rematching.Http.HttpManager;
import leapfrog_inc.rematching.System.Base64Utility;
import leapfrog_inc.rematching.System.Constants;

public class RoomRequester {

    public static class RoomData {

        public String id;
        public boolean approval;
        public int score;
        public int review;
        public String name;
        public String place;
        public int rent;
        public String phone;
        public String email;

        static public RoomData create(JSONObject json) {

            try {
                RoomData roomData = new RoomData();
                roomData.id = json.getString("id");
                roomData.approval = json.getString("approval").equals("1");
                roomData.score = Integer.parseInt(json.getString("score"));
                roomData.review = Integer.parseInt(json.getString("review"));
                roomData.name = Base64Utility.decode(json.getString("name"));
                roomData.place = Base64Utility.decode(json.getString("place"));
                roomData.rent = Integer.parseInt(json.getString("rent"));
                roomData.phone = Base64Utility.decode(json.getString("phone"));
                roomData.email = Base64Utility.decode(json.getString("email"));
                return roomData;

            } catch (Exception e) { }

            return null;
        }
    }

    private static RoomRequester mInstance = null;
    public ArrayList<RoomData> dataList = new ArrayList<RoomData>();

    private RoomRequester(){}

    public static RoomRequester getInstance(){
        if(mInstance == null){
            mInstance = new RoomRequester();
        }
        return mInstance;
    }

    public void fetch(final RoomRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject json = new JSONObject(data);
                        if (json.getString("result").equals("0")) {
                            ArrayList<RoomData> datas = new ArrayList<RoomData>();
                            JSONArray dataArray = json.getJSONArray("data");
                            for (int i = 0; i < dataArray.length(); i++) {
                                RoomData roomData = RoomData.create(dataArray.getJSONObject(i));
                                if (roomData != null) {
                                    datas.add(roomData);
                                }
                            }
                            dataList = datas;
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch (Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });

        StringBuffer param = new StringBuffer();
        param.append("command=getRoom");
        httpManager.execute(Constants.ServerApiUrl, "POST", param.toString());
    }

    public interface RoomRequesterCallback {
        void didReceiveData(boolean result);
    }

    public RoomData query(String roomId) {
        for (int i = 0; i < dataList.size(); i++) {
            RoomData roomData = dataList.get(i);
            if (roomData.id.equals(roomId)) {
                return roomData;
            }
        }
        return null;
    }
}
